### PR TITLE
#6 breakdown pt1 - add codegen

### PR DESCRIPTION
**Pull request by yvesfracari**
original PR: https://github.com/cowdao-grants/cow-py/pull/6

Added code generation capabilities
- Introduced web3_codegen in Makefile
- Updated codegen target to include Web3, orderbook, and subgraph codegen

Refactored project structure
- Removed contracts/ files (domain.py, order.py, sign.py)
- Eliminated order_book/ directory and contents
- Removed examples/ directory

Updated documentation
- Revised README.md for latest project structure
- Added note about signing orders in usage examples

Updated Chain enum for consistent explorer URL formatting
- Simplified imports in web3/provider.py
- Adjusted test suite, removing tests for refactored/removed components

### CODE REVIEW

**Review by ribeirojose — APPROVED** [review]
I wrote most of the commits here, so this is a bit weird 😆  but LGTM!

**Comment by gabrielfior** [comment]
I might be missing context, but I miss a bit of documentation here (in a README)
- How can one generate the autogen code
- Which package used
- When to update if needed

File: README.md [hunk 1]
```diff
@@ -141,6 +141,12 @@ cd cow-py
 poetry install
 ```
 
+If you need to interact with the contracts, please run the code generator. It should be updated if a new abi is added on the package.
+
+```bash
+python -m cow_py.codegen.main
```
Comment (ribeirojose) [inline README.md:147]: should be included in the Makefile?

File: Makefile [hunk 2]
```diff
@@ -2,8 +2,8 @@
 
 codegen: web3_codegen orderbook_codegen subgraph_codegen
 
-# web3_codegen:
-# 	poetry run web3_codegen
+web3_codegen:
+	poetry run python -m cow_py.codegen.main
```
Comment (ribeirojose) [inline Makefile:6]: 🙏

**Comment by ribeirojose** [comment]
> I might be missing context, but I miss a bit of documentation here (in a README)
> 
> * How can one generate the autogen code
> * Which package used
> * When to update if needed

I added more instructions here. Lmk if that looks good and I'll go ahead and merge this =)

File: cow_py/common/api/api_base.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    def __init__(self, config: APIConfig):
+        self.config = config
+
+    @with_backoff()
+    @rate_limitted()
+    async def _fetch(self, path, method="GET", **kwargs):
+        url = self.config.get_base_url() + path
+
+        del kwargs["context_override"]
```
Comment (gabrielfior) [inline cow_py/common/api/api_base.py:78]: Why is this needed?

File: tests/common/test_core_api.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import pytest
-from cow_py.common import constants, cow_error, chains, config
+
+from cow_py.common import chains, config, constants, cow_error
 
 
 @pytest.mark.parametrize("module", [constants, cow_error, chains, config])
```
Comment (gabrielfior) [inline tests/common/test_core_api.py:6]: I would like to see a real integration test (without mocked HTTP requests) which touch cowswap API, place an unrealistic order (e.g. buy ETH for 1 USD), cancel it and axit.
I apologize if such a test already exists (in that case, please close this comment), but I couldn't find it in the codebase from this PR.

Comment (yvesfracari) [reply tests/common/test_core_api.py:6]: On this PR, the API module didn't exist yet. This PR is just creating the codegen and adding some common utils. All API related logic and tests should be integrated here: https://github.com/cowdao-grants/cow-py/pull/13

Comment (gabrielfior) [reply tests/common/test_core_api.py:6]: This will be done on another PR
https://github.com/cowdao-grants/cow-py/pull/12/files#r1744309895
Closing this

File: cow_py/common/api/api_base.py (hunk 3 above)
Comment (ribeirojose) [reply cow_py/common/api/api_base.py]: added a comment explaining that, but this is essentially to avoid our backoff decorator args to be injected here
